Rewrite summary e2e test to check metric sanity

timstclair opened the pull request:
Take two, forked from https://github.com/kubernetes/kubernetes/pull/28195

Adds a test library that extends the ginkgo matchers to check nested data structures. Then uses the new matcher library to thoroughly check the validity of every field in the summary metrics API. This approach is more flexible than the previous approach since it allows for different tests per-field, and is easier to add case-by-case exceptions. It also places the lower & upper bounds side-by-side, making the test much easier to read & reason about.

Most fields are expected to be within some bounds. This is not intended to be a performance test, so metric bounds are very loose. Rather, I'm looking to check that the values are sane to catch bugs like #27194

Fixes #23411, https://github.com/kubernetes/kubernetes/issues/31989

/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29479)

<!-- Reviewable:end -->

timstclair commented:
Removed FIXME that was left in.

timstclair commented:
Hmm, still some kinks to work out. @vishh @ronnielai I'll reassign this once it's fixed.

timstclair commented:
Ok, I think this is (finally) ready for review again. I'm sorry it's so huge. The bulk of the new code is in the new matcher library, but when I tried to break that into a separate CL the feedback was that it would be better to keep it with the code that uses it. I'm also trying to get it upstreamed into ginkgo, but I suspect that will be a slow process. Let me know if there's anyway I can help ease the burden of this PR.

Also, I suspect this test will be a little, flaky at first while we tune the numbers more. I soaked it on my own VMs over the weekend, and didn't see any major issues, so hopefully it won't be too bad. The good news is the flakes should be easy to identify and fix :)

cc/ @spxtr

spxtr commented:
> Also, I suspect this test will be a little, flaky at first while we tune the numbers more. I soaked it on my own VMs over the weekend, and didn't see any major issues, so hopefully it won't be too bad. The good news is the flakes should be easy to identify and fix :)

That's worrying. Why do you suspect they'll be flaky?

timstclair commented:
Most of the bounds were found experimentally, by playing around with the test and seeing what ranges the numbers seemed to fall in, and then adding some padding (and verifying that the ranges look sane). However, I suspect there will be some outliers that none of my testing hit.

Is it possible to mark an individual test as none-critical? So that failures are logged (and an issue is filed), but it doesn't fail the build?

spxtr commented:
I think the closest thing we have to that is the `[Flaky]` label.

timstclair commented:
FYI, I just opened a gomega PR to upstream the matchers: https://github.com/onsi/gomega/pull/171

timstclair commented:
Actually, I think the gomega PR will beat this in, so I'll put this on hold to use the matchers from there instead.

spxtr commented:
Great!

tallclair committed:
Update gomega godeps to include gstruct

tallclair committed:
Rewrite summary stats test to validate metrics

timstclair commented:
Rebased with gomega upstreamed `gstruct` library (now you don't need to review as much :smile: )

@spxtr I added `[Flaky]` for now. I'll keep an eye on it once it's submitted, and if it seems stable enough in the flaky suite, move it out.

@Random-Liu This is ready for review (for real this time :) )

Random-Liu commented:
@timstclair Will review soon. :)

Random-Liu commented:
> @spxtr I added [Flaky] for now. I'll keep an eye on it once it's submitted, and if it seems stable enough in the flaky suite, move it out.

@timstclair IIRC, we don't have flaky suite for node e2e test.

timstclair commented:
Hmm.. What do you think about running it in the normal suite, and I'll promise to be on top of any flakes? If it turns out to be too flaky, we'll find another option.

Random-Liu commented:
@timstclair Hmm...Let me take a look at this PR first. ;)

timstclair commented:
Friendly ping :)

Random-Liu commented on test/e2e_node/summary_test.go:
```diff
@@ -0,0 +1,283 @@
+				}
+				return summary
+			}, 1*time.Minute, time.Second*15).Should(matchExpectations)
+		})
```
Add a Consistently check here?

timstclair replied:
Good idea. Done.

Random-Liu commented:
LGTM with one suggestion. And please rebase.

I'm not sure whether the bound is reasonable. However, since initially this will only run in flaky suite, we can observe the result and maybe ajust the bound if needed.

timstclair commented on test/e2e_node/summary_test.go:
```diff
@@ -0,0 +1,283 @@
+
+			// Setup expectations.
+			const (
+				kb = 1000
```
I think it would be `kib` for 1024. In practice it doesn't matter, because the bounds aren't that precise.

Random-Liu commented:
LGTM

timstclair commented:
Thanks! Squashed, self-applying LGTM.

timstclair commented:
Reran `hack/update-godep-licenses.sh`, re-applying LGTM.

timstclair commented:
@k8s-bot e2e test this, issue: https://github.com/kubernetes/kubernetes/issues/33388

Random-Liu commented:
@timstclair Have you run it before? It failed on my machine with a lot of errors, such as:
```
 Expected
        <string>: Summary
    to match fields: {
    [[[.Node.SystemContainers[kubelet].StartTime:
    	Expected a time.Time.  Got:
    	    <unversioned.Time>: {
    	        Time: {
...
```

Random-Liu commented:
Another thing is that can we add the original summary test back until we move the new test out of flaky suite?

ixdy commented:
This PR broke the linux/arm crossbuild:
```
+++ [0928 20:02:01] linux/arm: go build started
# k8s.io/kubernetes/test/e2e_node
test/e2e_node/summary_test.go:57: constant 100000000000 overflows int
test/e2e_node/summary_test.go:157: constant 100000000000 overflows int
test/e2e_node/summary_test.go:158: constant 10000000000 overflows int
test/e2e_node/summary_test.go:159: constant 10000000000 overflows int
test/e2e_node/summary_test.go:167: constant 100000000000 overflows int
test/e2e_node/summary_test.go:169: constant 10000000000 overflows int
test/e2e_node/summary_test.go:175: constant 10000000000 overflows int
test/e2e_node/summary_test.go:183: constant 10000000000 overflows int
```

I think nobody noticed yet because Jenkins is out of email quota for the month.
cc @luxas

luxas commented:
I think the int should just have the specific int64 type...
Now int => int64 on all other platforms, but => int32 on arm, which makes it fail.

@timstclair Can you change the constant to type int64 please?

timstclair commented:
I'll fix the arm build & add the old test back.

@Random-Liu - Yes, it passes on my local machine. Can you post the full output when you run locally to a gist?
